test/csuite/wt3363_checkpoint_op_races/main.c

```diff
@@ -0,0 +1,498 @@
+	testutil_check(__wt_snprintf(config, sizeof(config),
+	    "force,debug_checkpoint_latency=%" PRIu64, max_execution_time));
+	(void)time(&start);
+	(void)time(&now);
+
+	while (difftime(now, start) < RUNTIME) {
+		if ((ret = conn->open_session(conn, NULL, NULL, &session)) != 0)
+			testutil_die(ret, "conn.session");
+
+		if ((ret = session->checkpoint(session, config)) != 0)
+			if (ret != EBUSY && ret != ENOENT)
+				testutil_die(ret, "session.checkpoint");
+
+		if ((ret = session->close(session, NULL)) != 0)
+			testutil_die(ret, "session.close");
+		(void)time(&now);
+
+		/*
+		 * A short sleep to let operations process and avoid back to
+		 * back checkpoints locking up resources.
+		 */
+		sleep(1);
+	}
+
+	return (NULL);
+}
+
+/*
+ * Function to monitor running operations and abort to dump core in the event
+ * that we catch an operation running long.
+ */
+void *
+monitor(void *args)
+{
+	time_t now, start;
+	uint64_t last_ops[N_THREADS];
+	int i;
+
+	/* Unused */
+	(void)args;
+
+	(void)time(&start);
+	(void)time(&now);
+
+	for (i = 0; i < N_THREADS; i++)
+		last_ops[i] = 0;
+
+	while (difftime(now, start) < RUNTIME) {
+		/*
+		 * Checkpoints will run for slightly over max_execution_time.
+		 * max_execution_times should always be long enough that we can
+		 * complete any single operation in 1/2 that time.
+		 */
+		sleep(max_execution_time/2);
+
+		for (i = 0; i < N_THREADS; i++) {
+			/* Ignore any threads which may not have started yet */
+			if (thread_counters[i] == 0)
+				continue;
+			/*
+			 * We track how many operations each thread has done. If
+			 * we have slept and the counter remains the same for a
+			 * thread it is stuck and should drop a core so the
+			 * cause of the hang can be investigated.
+			 */
+			if (thread_counters[i] != last_ops[i])
+				last_ops[i] = thread_counters[i];
+			else {
+				printf("Thread %d had a task running"
+				    " for more than %" PRIu64 " seconds\n",
+				    i, max_execution_time/2);
+				abort();
+
+			}
+		}
+		(void)time(&now);
+	}
+
+	return (NULL);
+}
+
+/* Worker thread. Executes random operations from the set of 6 */
+void *
+do_ops(void *args)
+{
+	THREAD_ARGS *arg;
+	WT_CONNECTION *conn;
+	WT_RAND_STATE rnd;
+	time_t now, start;
+
+	arg = (THREAD_ARGS *)args;
+	conn = arg->testopts->conn;
+	__wt_random_init_seed(NULL, &rnd);
+	(void)time(&start);
+	(void)time(&now);
+
+	while (difftime(now, start) < RUNTIME) {
+		switch (__wt_random(&rnd) % 6) {
+			case 0:
+				op_bulk(conn);
+				break;
+			case 1:
+				op_create(conn);
+				break;
+			case 2:
+				op_cursor(conn);
+				break;
+			case 3:
+				op_drop(conn, __wt_random(&rnd) & 1);
+				break;
+			case 4:
+				op_bulk_unique(conn,
+				    __wt_random(&rnd) & 1, arg->threadnum);
+				break;
+			case 5:
+				op_create_unique(conn,
+				    __wt_random(&rnd) & 1, arg->threadnum);
+				break;
+		}
+		/* Increment how many ops this thread has performed */
+		__wt_atomic_add64(&thread_counters[arg->threadnum], 1);
```

Comment:
Why use atomic add here, but not read atomically? I think this code will change around when you take my suggestion about using the THREAD_ARGS instead of a global.